### ISSUE TITLE
[FEATURE] dRAID expansion

### ISSUE THREAD

**Issue by Sawtaytoes** [openzfs/zfs#15413]
### Describe the feature would like to see added to OpenZFS

I'd like to see the ability to expand dRAID vdevs. Since they used fixed-size stripes and since RAID-Z expansion is coming up soon, I'd like to see the same feature available for dRAID.

### How will this feature improve OpenZFS?

dRAID is meant to contain hundreds of drives. Unlike RAID-Z, where you typically want smaller vdevs, dRAID is designed for scale.

Since dRAID acts like its own zpool, you'll wanna throw a ton of drives at it. Let's say you have 100 drives in a single dRAID. If you want to expand this pool, the easiest way is to drop 10-50 more drives into the current dRAID.

But without dRAID expansion, you'll need to either add a separate dRAID or hold onto those drives until you have another 100 (or enough that you'd make another dRAID). Or worse yet, you'd need to get 200 drives, create a new pool with a 200-drive dRAID, and then move all your data over, then swap the pools without interrupting users and trash the existing 100 drives.

The long-term scalability of a single dRAID vdev makes it perfect for expansion.

### Additional context

Is it possible to use the existing RAID-Z expansion code for dRAID? From what I've seen, as soon as you expand a RAID-Z, it gets converted in a way that's very similar to dRAID.

**Comment by amotin** [comment]
dRAID expansion would likely be completely different from RAIDZ expansion.  While the last changes stripe width, the first would only reshuffle the data according to the new number of drives in vdev, not touching the stripe size.  I'd actually expect it to be easier and cleaner at the end, but it would be a completely different process.  Though I doubt in demand for such expansion. People using dRAID are not usually inserting 10 drives, they are inserting full JBODs and racks at once.

**Comment by Sawtaytoes** [comment]
Thanks for that clarification!

### Enterprise perspective
I guess if I bought an entire JBOD, I'd setup a separate dRAID for that rack and add the vdev to my existing zpool if I needed more storage capacity rather than adding new drives.

### Home user perspective
Maybe my use case is unusual since I'm still using hundreds of smaller 2-4TB SSDs.

A typical home user might simply have a bunch of absurdly large HDDs with 1-2 RAID-Z2 vdevs maxing around 15-30 drives. With RAID-Z expansion already on the way, it's simple to add a new vdev and scale up as necessary.

## dRAID can (mostly) replace RAID-Z
I see dRAID as a RAID-Z replacement but has some drawbacks:
1. dRAID practically requires a metadata (special) vdev; although, that's good with any HDD deployment. I even use metadata Optane devices to speed up NAND SSD latency.
2. It doesn't utilize space efficiently with small files <32kB. I used to setup Windows with 64kB blocks back when SSDs first came to the market, and I don't think 32kB blocks are gonna be that much worse. Git repos make this problematic though.

Even for smaller deployments, I think home users will be tempted to setup dRAID because it's much simpler than RAID-Z and fixes the resilvering problem making it a RAID-Z configuration for people like me that use all-mirrors.

**Comment by Sawtaytoes** [comment]
Is it possible to donate, so I could sponsor this feature?

What's something like this cost? $1K? $10K? $100K?

**Comment by hachre** [comment]
I'd like to see this as well. In my testing DRAID far outperforms RAIDZ2 when it comes to rebuilding an array. However as a small scale user the same needs apply as those that we already expressed regarding RAIDZ expansion: the ability to start with few disks and scale up over time is critical for use in a small scale scenario. DRAID has clear advantages over RAIDZ once you reach a certain disk number, but you are not going to start with that disk number outright. That's the general difference between small scale users and enterprise users.

**Comment by RSully** [comment]
> Even for smaller deployments, I think home users will be tempted to setup dRAID because it's much simpler for more than 10 drives than RAID-Z and fixes the resilvering problem making it a RAID-Z configuration for people like me that use all-mirrors

Absolutely. I am putting together a new system and am planning to start with around 8-10 disks. Something like a draid2:4d:2s pool would be a good balance of risk to usable capacity for me, **if** I knew I could expand it down the line. Instead, to avoid RAIDZ2 resilvering times I am likely going to go with 3x 3-wide mirrors and a hot spare.
